public/app/features/explore/AddToDashboard/AddToDashboardModal.tsx

```diff
@@ -1,136 +1,155 @@
-import { FolderPicker } from 'app/core/components/Select/FolderPicker';
-import { useForm } from 'react-hook-form';
-import { SaveToNewDashboardDTO } from './addToDashboard';
-
-export interface ErrorResponse {
-  status: string;
-  message?: string;
+import React from 'react';
+import { Button, Field, InputControl, Modal, RadioButtonGroup, useStyles2 } from '@grafana/ui';
+import { SelectableValue } from '@grafana/data';
+import { addPanelToDashboard } from './addToDashboard';
+import { useSelector } from 'react-redux';
+import { ExploreId } from 'app/types';
+import { DashboardPicker } from 'app/core/components/Select/DashboardPicker';
+import { DeepMap, FieldError, useForm } from 'react-hook-form';
+import { css } from '@emotion/css';
+import { config, locationService } from '@grafana/runtime';
+import { getExploreItemSelector } from '../state/selectors';
+
+enum SaveTarget {
+  NewDashboard,
+  ExistingDashboard,
 }
 
-const ERRORS = {
-  NAME_REQUIRED: 'Dashboard name is required.',
-  NAME_EXISTS: 'A dashboard with the same name already exists in this folder.',
-  INVALID_FIELD: 'This field is invalid.',
-  UNKNOWN_ERROR: 'An unknown error occurred while saving the dashboard. Please try again.',
-  INVALID_FOLDER: 'Select a valid folder to save your dashboard in.',
-};
+const SAVE_TARGETS: Array<SelectableValue<SaveTarget>> = [
+  {
+    label: 'New Dashboard',
+    value: SaveTarget.NewDashboard,
+  },
+  {
+    label: 'Existing Dashboard',
+    value: SaveTarget.ExistingDashboard,
+  },
+];
 
-type FormDTO = SaveToNewDashboardDTO;
+interface SaveTargetDTO {
+  saveTarget: SaveTarget;
+}
+interface SaveToNewDashboardDTO extends SaveTargetDTO {
+  saveTarget: SaveTarget.NewDashboard;
+}
 
-interface Props {
-  onClose: () => void;
-  onSave: (data: FormDTO, redirect: boolean) => Promise<void | ErrorResponse>;
+interface SaveToExistingDashboard extends SaveTargetDTO {
+  saveTarget: SaveTarget.ExistingDashboard;
+  dashboardUid: string;
+}
+
+type FormDTO = SaveToNewDashboardDTO | SaveToExistingDashboard;
+
+function assertIsSaveToExistingDashboardError(
+  errors: DeepMap<FormDTO, FieldError>
+): asserts errors is DeepMap<SaveToExistingDashboard, FieldError> {
+  // the shape of the errors object is always compatible with the type above, but we need to
+  // explicitly assert its type so that TS can narrow down FormDTO to SaveToExistingDashboard
+  // when we use it in the form.
 }
 
 function withRedirect<T extends any[]>(fn: (redirect: boolean, ...args: T) => {}, redirect: boolean) {
   return async (...args: T) => fn(redirect, ...args);
 }
 
-export const AddToDashboardModal = ({ onClose, onSave }: Props) => {
-  const [submissionError, setSubmissionError] = useState<string>();
+function openDashboard(openInNewTab: boolean, dashboardUid?: string) {
+  const url = dashboardUid ? `d/${dashboardUid}` : 'dashboard/new';
+
+  if (!openInNewTab) {
+    locationService.push(url);
+  } else {
+    global.open(config.appUrl + url, '_blank');
```

Comment:
I think we need better error handling here as well. If a user uses a browser plugin blocking new windows we will end up with a panel in local storage that is not automatically picked up. What's worse such a stale panel would be added to a newly added dashboard. 

Alternatively, we could use `postMessage`, but that could be improved later on as long as we have good error handling. Using local storage to pass data between tabs is a bit clunky because it requires both saving and reading to work in a transaction-like flow. Also, it simply won't work with local storage disabled or full.

Reply:
We use local storage for many many features. (Rich query history among others)

But yes some error handling would be good 👍

Reply:
added some error handling in 9e15845379 (probably needs some better wording):

localStorage related errors:
![Screenshot 2022-04-06 at 11 33 39](https://user-images.githubusercontent.com/1170767/161956090-b60a01c6-0c7d-4e50-92d9-8fe6a06a2ce0.png)

when fetching the selected dashboard fails:
![Screenshot 2022-04-06 at 11 34 19](https://user-images.githubusercontent.com/1170767/161956195-8f802691-e78a-4309-92c8-8f149dff5d2c.png)

opening a new tab fails:
![Screenshot 2022-04-06 at 11 35 04](https://user-images.githubusercontent.com/1170767/161956303-e77f83b7-1969-4ed8-b58f-67e1192387f4.png)

Reply:
Noice!

Reply:
added a bunch of tests for this in https://github.com/grafana/grafana/pull/47083/commits/d4a8b78c7792ea134edcbad81d8bd01532b55d81